src/azul/queues.py

```diff
@@ -86,13 +90,30 @@
 @attrs.frozen(kw_only=True)
 class SQSMessage:
     body: JSON
+    # These fields are None when the message hasn't been sent yet, and not None
+    # after it's been received
+    attempts: int | None = None
+    id: str | None = None
 
     def to_entry(self) -> 'SendMessageRequestQueueSendMessageTypeDef':
         return {'MessageBody': json.dumps(self.body)}
 
     def to_batch_entry(self, id: int) -> 'SendMessageBatchRequestEntryTypeDef':
         return {**self.to_entry(), 'Id': str(id)}
 
+    @classmethod
+    def from_record(cls, record: SQSRecord) -> Self:
+        record = record.to_dict()
+        return cls(**cls._extract_fields(record, json_mapping(record['attributes'])))
+
+    @classmethod
+    def _extract_fields(cls, record: JSON, attributes: JSON) -> dict[str, Any]:
+        return {
+            'id': json_str(record['messageId']),
```

Comment:
You shouldn't have to narrow the type, only to then discard that constraint again via the wide return type. I understand that we do want to narrow somewhere, so maybe this is a bit too much, and we should inline this method, get stronger static type safety and live with the duplication.